util/gitutil/git_url.go

```diff
@@ -38,40 +54,93 @@ var protoRegexp = regexp.MustCompile(`^[a-zA-Z0-9]+://`)
 //
 // See https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols
 type GitURL struct {
-	// Scheme is the protocol over which the git repo can be accessed
-	Scheme string
-
-	// Host is the remote host that hosts the git repo
-	Host string
-	// Path is the path on the host to access the repo
-	Path string
+	GitURLBase
 	// User is the username/password to access the host
 	User *url.Userinfo
 	// Opts can contain additional metadata
 	Opts *GitURLOpts
-
-	// Remote is a valid URL remote to pass into the Git CLI tooling (i.e.
-	// without the fragment metadata)
-	Remote string
 }
 
 // GitURLOpts is the buildkit-specific metadata extracted from the fragment
-// of a remote URL.
+// or the query of a remote URL.
 type GitURLOpts struct {
 	// Ref is the git reference
 	Ref string
+	// Checksum is the commit hash
+	Checksum string
 	// Subdir is the sub-directory inside the git repository to use
 	Subdir string
 }
 
+// GitURLOptsError is returned for invalid GitURLOpts.
+type GitURLOptsError struct {
+	error
+}
+
 // parseOpts splits a git URL fragment into its respective git
 // reference and subdirectory components.
-func parseOpts(fragment string) *GitURLOpts {
-	if fragment == "" {
-		return nil
+func parseOpts(fragment string, query url.Values) (*GitURLOpts, error) {
```

Comment:
Querystring is only part of the new dockerfile-style URLs so I don't understand why this parsing is in this pacakge.